File: packages/rune-ios/ios/RuneKit/src/utils/RuneGradientParser.m
#import "RuneGradientParser.h"
#import "RuneColorParser.h"
#import <math.h>

static NSArray<NSString *> *SNRuneSplitGradientArgs(NSString *raw) {
  NSMutableArray<NSString *> *parts = [NSMutableArray array];
  NSMutableString *current = [NSMutableString string];
  NSInteger depth = 0;
  for (NSUInteger i = 0; i < raw.length; i++) {
    unichar c = [raw characterAtIndex:i];
    if (c == ',') {
      if (depth == 0) {
        NSString *trimmed = [current stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
        if (trimmed.length) [parts addObject:trimmed];
        [current setString:@""];
      } else {
        [current appendFormat:@"%C", c];
      }
    } else if (c == '(') {
      depth++;
      [current appendFormat:@"%C", c];
    } else if (c == ')') {
      if (depth > 0) depth--;
      [current appendFormat:@"%C", c];
    } else {
      [current appendFormat:@"%C", c];
    }
  }
  NSString *final = [current stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
  if (final.length) [parts addObject:final];
  return parts;
}

static CGFloat SNRuneNormalizeAngle(CGFloat angle) {
  CGFloat a = fmod(angle, 360.0);
  if (a < 0) a += 360.0;
  return a;
}

static CGFloat SNRuneParseDirection(NSString *dir, BOOL *ok) {
  NSArray<NSString *> *parts = [[dir lowercaseString] componentsSeparatedByCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
  NSNumber *vertical = nil;
  NSNumber *horizontal = nil;
  for (NSString *p in parts) {
    if (!p.length) continue;
    if ([p isEqualToString:@"top"]) vertical = @(0.f);
    else if ([p isEqualToString:@"bottom"]) vertical = @(180.f);
    else if ([p isEqualToString:@"left"]) horizontal = @(270.f);
    else if ([p isEqualToString:@"right"]) horizontal = @(90.f);
  }
  if (vertical && horizontal) {
    *ok = YES;
    if (vertical.floatValue == 0.f && horizontal.floatValue == 90.f) return 45.f;
    if (vertical.floatValue == 0.f && horizontal.floatValue == 270.f) return 315.f;
    if (vertical.floatValue == 180.f && horizontal.floatValue == 90.f) return 135.f;
    if (vertical.floatValue == 180.f && horizontal.floatValue == 270.f) return 225.f;
    return vertical.floatValue;
  }
  if (vertical) { *ok = YES; return vertical.floatValue; }
  if (horizontal) { *ok = YES; return horizontal.floatValue; }
  *ok = NO;
  return 0.f;
}

static CGFloat SNRuneParseAngleToken(NSString *token, BOOL *ok) {
  NSString *trimmed = [token stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
  if ([trimmed hasPrefix:@"to "]) {
    BOOL dirOk = NO;
    CGFloat angle = SNRuneParseDirection([trimmed substringFromIndex:3], &dirOk);
    *ok = dirOk;
    return angle;
  }

  NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:@"^(-?[0-9.]+)(deg|rad|turn)?$" options:NSRegularExpressionCaseInsensitive error:nil];
  NSTextCheckingResult *match = [regex firstMatchInString:trimmed options:0 range:NSMakeRange(0, trimmed.length)];
  if (!match) { *ok = NO; return 0.f; }

  NSString *valueString = [trimmed substringWithRange:[match rangeAtIndex:1]];
  NSString *unit = match.numberOfRanges > 2 ? [[trimmed substringWithRange:[match rangeAtIndex:2]] lowercaseString] : @"";
  CGFloat value = valueString.doubleValue;
  *ok = YES;
  if (unit.length == 0 || [unit isEqualToString:@"deg"]) return value;
  if ([unit isEqualToString:@"rad"]) return (CGFloat)(value * 180.0 / M_PI);
  if ([unit isEqualToString:@"turn"]) return (CGFloat)(value * 360.0);
  *ok = NO;
  return 0.f;
}

static NSNumber *_Nullable SNRuneParsePosition(NSString *raw) {
  NSString *trimmed = [raw stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
  if ([trimmed hasSuffix:@"%"]) {
    CGFloat v = [[trimmed substringToIndex:trimmed.length - 1] doubleValue] / 100.0;
    v = fmax(0.f, fmin(1.f, v));
    return @(v);
  }
  double v = trimmed.doubleValue;
  if (isnan(v)) return nil;
  if (fabs(v) > 1.0) v = v / 100.0;
  v = fmax(0.f, fmin(1.f, v));
  return @(v);
}

static RuneGradientStop *_Nullable SNRuneParseStop(NSString *token) {
  NSString *trimmed = [token stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
  if (!trimmed.length) return nil;

  NSInteger depth = 0;
  NSInteger splitIndex = -1;
  for (NSUInteger i = 0; i < trimmed.length; i++) {
    unichar c = [trimmed characterAtIndex:i];
    if (c == '(') depth++;
    else if (c == ')' && depth > 0) depth--;
    else if ((c == ' ' || c == '\t') && depth == 0) { splitIndex = (NSInteger)i; break; }
  }

  NSString *colorPart = splitIndex == -1 ? trimmed : [trimmed substringToIndex:(NSUInteger)splitIndex];
  NSString *positionPart = splitIndex == -1 ? nil : [trimmed substringFromIndex:(NSUInteger)splitIndex];

  UIColor *color = [RuneColorParser parseColor:[colorPart stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]]];
  if (!color) return nil;
  NSNumber *pos = positionPart ? SNRuneParsePosition(positionPart) : nil;
  return [[RuneGradientStop alloc] initWithColor:color position:pos];
}

@implementation RuneGradientStop

- (instancetype)initWithColor:(UIColor *)color position:(NSNumber *)position {
  if (self = [super init]) {
    _color = color;
    _position = position;
  }
  return self;
}

@end

@implementation RuneLinearGradient

- (instancetype)initWithAngle:(CGFloat)angle stops:(NSArray<RuneGradientStop *> *)stops {
  if (self = [super init]) {
    _angle = angle;
    _stops = stops;
  }
  return self;
}

@end

@implementation RuneGradientParser

+ (nullable RuneLinearGradient *)parse:(id)value {
  NSString *raw = nil;
  id candidate = value;
  if ([candidate isKindOfClass:[NSString class]]) {
    NSString *trimmed = [(NSString *)candidate stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
    if ([trimmed hasPrefix:@"["]) {
      NSData *data = [trimmed dataUsingEncoding:NSUTF8StringEncoding];
      id json = [NSJSONSerialization JSONObjectWithData:data options:0 error:nil];
      if ([json isKindOfClass:[NSArray class]]) {
        candidate = json;
      }
    }
  }

  if ([candidate isKindOfClass:[NSArray class]]) {
    id first = [(NSArray *)candidate firstObject];
    if ([first isKindOfClass:[NSString class]]) raw = (NSString *)first;
  } else if ([candidate isKindOfClass:[NSString class]]) {
    raw = (NSString *)candidate;
  } else if ([candidate respondsToSelector:@selector(description)]) {
    raw = [[candidate description] copy];
  }
  if (raw.length == 0) return nil;
  NSString *trimmed = [raw stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
  if (![trimmed length] || ![[trimmed lowercaseString] hasPrefix:@"linear-gradient"]) return nil;

  NSRange open = [trimmed rangeOfString:@"("];
  NSRange close = [trimmed rangeOfString:@")" options:NSBackwardsSearch];
  if (open.location == NSNotFound || close.location == NSNotFound || close.location <= open.location) return nil;
  NSString *inside = [trimmed substringWithRange:NSMakeRange(open.location + 1, close.location - open.location - 1)];
  NSArray<NSString *> *tokens = SNRuneSplitGradientArgs(inside);
  if (tokens.count < 2) return nil;

  CGFloat angle = 180.f;
  NSUInteger startIndex = 0;
  BOOL angleOk = NO;
  angle = SNRuneParseAngleToken(tokens.firstObject, &angleOk);
  if (angleOk) {
    startIndex = 1;
  } else {
    angle = 180.f;
  }

  NSMutableArray<RuneGradientStop *> *stops = [NSMutableArray array];
  for (NSUInteger i = startIndex; i < tokens.count; i++) {
    RuneGradientStop *stop = SNRuneParseStop(tokens[i]);
    if (stop) [stops addObject:stop];
  }
  if (stops.count < 2) return nil;

  return [[RuneLinearGradient alloc] initWithAngle:SNRuneNormalizeAngle(angle) stops:stops];
}

@end

File: packages/rune-android/android/RuneKit/src/main/java/com/rune/kit/core/RuneGradientParser.kt
package com.rune.kit.core

import org.json.JSONArray
import kotlin.math.PI
import kotlin.math.abs

data class RuneGradientStop(val color: Int, val position: Float?)
data class RuneLinearGradient(val angle: Float, val stops: List<RuneGradientStop>)

object RuneGradientParser {

  fun parse(value: Any?): RuneLinearGradient? {
    var candidate: Any? = value
    if (candidate is String) {
      val trimmedStr = candidate.trim()
      if (trimmedStr.startsWith("[")) {
        runCatching { JSONArray(trimmedStr) }.getOrNull()?.let { candidate = it }
      } else {
        candidate = trimmedStr
      }
    }
    val raw = when (val c = candidate) {
      is JSONArray -> if (c.length() > 0) c.optString(0) else null
      is String -> c
      else -> c?.toString()
    }?.trim() ?: return null

    return parseLinearGradient(raw)
  }

  private fun parseLinearGradient(raw: String): RuneLinearGradient? {
    if (!raw.startsWith("linear-gradient", ignoreCase = true)) return null
    val open = raw.indexOf('(')
    val close = raw.lastIndexOf(')')
    if (open < 0 || close <= open) return null
    val args = raw.substring(open + 1, close)
    val tokens = splitArgs(args)
    if (tokens.size < 2) return null

    var angle = 180f
    var startIndex = 0
    parseAngle(tokens.first())?.let {
      angle = it
      startIndex = 1
    }

    val stops = mutableListOf<RuneGradientStop>()
    for (i in startIndex until tokens.size) {
      parseStop(tokens[i])?.let { stops.add(it) }
    }
    if (stops.size < 2) return null

    return RuneLinearGradient(normalizeAngle(angle), stops)
  }

  private fun parseAngle(token: String): Float? {
    val trimmed = token.trim()
    if (trimmed.startsWith("to ", ignoreCase = true)) {
      return parseDirection(trimmed.removePrefix("to ").trim())
    }
    val pattern = Regex("^(-?[0-9.]+)(deg|rad|turn)?$", RegexOption.IGNORE_CASE)
    val match = pattern.find(trimmed) ?: return null
    val value = match.groupValues[1].toFloatOrNull() ?: return null
    return when (match.groupValues[2].lowercase()) {
      "deg", "" -> value
      "rad" -> (value * 180f / PI.toFloat())
      "turn" -> value * 360f
      else -> null
    }
  }

  private fun parseDirection(dir: String): Float? {
    val parts = dir.lowercase().split(Regex("\\s+")).filter { it.isNotBlank() }
    var vertical: Float? = null
    var horizontal: Float? = null
    for (p in parts) {
      when (p) {
        "top" -> vertical = 0f
        "bottom" -> vertical = 180f
        "left" -> horizontal = 270f
        "right" -> horizontal = 90f
      }
    }
    return when {
      vertical != null && horizontal != null -> {
        when {
          vertical == 0f && horizontal == 90f -> 45f
          vertical == 0f && horizontal == 270f -> 315f
          vertical == 180f && horizontal == 90f -> 135f
          vertical == 180f && horizontal == 270f -> 225f
          else -> vertical ?: horizontal
        }
      }
      vertical != null -> vertical
      horizontal != null -> horizontal
      else -> null
    }
  }

  private fun parseStop(token: String): RuneGradientStop? {
    val trimmed = token.trim()
    if (trimmed.isEmpty()) return null

    var splitIndex = -1
    var depth = 0
    for (i in trimmed.indices) {
      when (val c = trimmed[i]) {
        '(' -> depth++
        ')' -> if (depth > 0) depth--
        ' ', '\t' -> if (depth == 0) { splitIndex = i; break }
      }
    }

    val colorPart = if (splitIndex == -1) trimmed else trimmed.substring(0, splitIndex).trim()
    val positionPart = if (splitIndex == -1) null else trimmed.substring(splitIndex).trim()

    val color = RuneColorParser.parse(colorPart) ?: return null
    val position = positionPart?.let { parsePosition(it) }
    return RuneGradientStop(color, position)
  }

  private fun parsePosition(raw: String): Float? {
    val trimmed = raw.trim()
    if (trimmed.endsWith("%")) {
      val num = trimmed.removeSuffix("%").toFloatOrNull() ?: return null
      return (num / 100f).coerceIn(0f, 1f)
    }
    val num = trimmed.toFloatOrNull() ?: return null
    return if (abs(num) > 1f) {
      (num / 100f).coerceIn(0f, 1f)
    } else {
      num.coerceIn(0f, 1f)
    }
  }

  private fun splitArgs(raw: String): List<String> {
    val parts = mutableListOf<String>()
    val current = StringBuilder()
    var depth = 0
    for (c in raw) {
      when (c) {
        ',' -> {
          if (depth == 0) {
            parts.add(current.toString().trim())
            current.clear()
          } else {
            current.append(c)
          }
        }
        '(' -> {
          depth++
          current.append(c)
        }
        ')' -> {
          if (depth > 0) depth--
          current.append(c)
        }
        else -> current.append(c)
      }
    }
    if (current.isNotEmpty()) {
      parts.add(current.toString().trim())
    }
    return parts.filter { it.isNotEmpty() }
  }

  private fun normalizeAngle(angle: Float): Float {
    var a = angle % 360f
    if (a < 0f) a += 360f
    return a
  }
}
